File: packages/sdk/src/rules/schema-validator.ts
import { createRequire } from 'node:module';
import { POLICY_IR_V1_SCHEMA } from './policy-ir-schema.js';

const require = createRequire(import.meta.url);

export class PolicySchemaError extends Error {
  readonly errors: PolicyValidationError[];

  constructor(errors: PolicyValidationError[]) {
    const summary = errors
      .map((e) => `  - ${e.path}: ${e.message}`)
      .join('\n');
    super(`Invalid policy document:\n${summary}`);
    this.name = 'PolicySchemaError';
    this.errors = errors;
  }
}

export interface PolicyValidationError {
  path: string;
  message: string;
  keyword: string;
}

interface AjvErrorObject {
  instancePath: string;
  message?: string;
  keyword: string;
}

interface ValidateFunction {
  (data: unknown): boolean;
  errors?: AjvErrorObject[] | null;
}

interface AjvInstance {
  compile(schema: object): ValidateFunction;
}

interface AjvConstructor {
  new (opts: { allErrors: boolean }): AjvInstance;
}

let _validate: ValidateFunction | null = null;

function getValidator(): ValidateFunction {
  if (!_validate) {
     
    const { Ajv2020 } = require('ajv/dist/2020') as { Ajv2020: AjvConstructor };
    const ajv = new Ajv2020({ allErrors: true });
    _validate = ajv.compile(POLICY_IR_V1_SCHEMA);
  }
  return _validate;
}

function formatErrors(errors: AjvErrorObject[]): PolicyValidationError[] {
  return errors.map((err) => ({
    path: err.instancePath || '/',
    message: err.message ?? `failed ${err.keyword} validation`,
    keyword: err.keyword,
  }));
}

export function validatePolicyIR(data: unknown): void {
  const validate = getValidator();
  const valid = validate(data);
  if (!valid && validate.errors) {
    throw new PolicySchemaError(formatErrors(validate.errors));
  }
}

File: packages/sdk-python/veto/rules/schema_validator.py
"""
Policy IR v1 schema validator.

Validates policy documents against the canonical JSON Schema
shared between TypeScript and Python SDKs.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from jsonschema import Draft202012Validator, ValidationError as JsonSchemaError

from veto.rules.policy_ir_schema import POLICY_IR_V1_SCHEMA


@dataclass
class PolicyValidationError:
    """A single validation error with location and message."""

    path: str
    message: str
    keyword: str


class PolicySchemaError(Exception):
    """Raised when a policy document fails schema validation."""

    def __init__(self, errors: list[PolicyValidationError]) -> None:
        summary = "\n".join(f"  - {e.path}: {e.message}" for e in errors)
        super().__init__(f"Invalid policy document:\n{summary}")
        self.errors = errors


_validator: Draft202012Validator | None = None


def _get_validator() -> Draft202012Validator:
    global _validator
    if _validator is None:
        Draft202012Validator.check_schema(POLICY_IR_V1_SCHEMA)
        _validator = Draft202012Validator(POLICY_IR_V1_SCHEMA)
    return _validator


def _format_path(error: JsonSchemaError) -> str:
    parts = [str(p) for p in error.absolute_path]
    return "/" + "/".join(parts) if parts else "/"


def validate_policy_ir(data: Any) -> None:
    """Validate a policy document against the Policy IR v1 schema.

    Args:
        data: Parsed policy document (dict from YAML/JSON).

    Raises:
        PolicySchemaError: If validation fails, with actionable error details.
    """
    validator = _get_validator()
    raw_errors = list(validator.iter_errors(data))
    if raw_errors:
        errors = [
            PolicyValidationError(
                path=_format_path(e),
                message=e.message,
                keyword=e.validator,
            )
            for e in raw_errors
        ]
        raise PolicySchemaError(errors)
